### ISSUE TITLE
XSD for Version 3 output format

### ISSUE THREAD

**Issue by oznetmaster** [nunit/nunit#281]
Is there an XSD for the 3.0 output format?  If not, should there be one?

**Comment by ravichandranjv** [comment]
Unable to attach file for some reason.

The suggestion to use a tool to generate the XSD from the XML file is not so simple because there are some restrictions that need to be taken care of in the output XML file as specifying only certain values to be present for some elements.

@CharliePoole The "Description" value is missing in the TestResult format described in the github so the below XSD is still not complete.. 

Plus, there are many other maintenance related points like global or local change impact of the schema in future that may impact the XML validation, which should be discussed here. 

This schema I have tested for only two elements - **TestRun** (test-run) and **Environment** but still would like someone to test if the schema validates correctly.

Assumption: the root element is named TestRun without the '-' as in the below XML, which is a modified file of the actual TestResults.xml. 

**The XML**

<?xml version="1.0" encoding="utf-8" standalone="no"?>
<TestRun xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.nunit.org testresults.xsd">
id="2" testcasecount="2" total="2" passed="2" failed="0" inconclusive="0" skipped="0" asserts="0" start-time="2015-10-07 13:30:06Z" end-time="2015-10-07 13:30:07Z" >
<environment nunit-version="3.0.5715.30860" clr-version="4.0.30319.42000" os-version="Microsoft Windows NT 6.1.7601 Service Pack 1" platform="Win32NT" cwd="E:\nunit3_tests" machine-name="JV-PC" user="Jv" user-domain="Jv-PC" culture="en-US" uiculture="en-US" os-architecture="x86" />
</TestRun>

**The XSD**

<?xml version="1.0"?>
<xsd:schema xmlns:xsd="http://www.w3.org/2001/XMLSchema"
 xmlns:nunit="http://www.nunit.org"
 elementFormDefault="unqualified"
 attributeFormDefault="unqualified">
<xsd:element name="TestRun" />
 <xsd:complexType name="etr">
	<xsd:sequence>
	 <xsd:element ref="Environment" minOccurs="1" maxOccurs="unbounded" />
	 <xsd:element ref="TestSuite" minOccurs="1" maxOccurs="unbounded" />
	 <xsd:element ref="CommandLine" minOccurs="1" maxOccurs="unbounded" />
	 <xsd:element ref="Settings" minOccurs="1" maxOccurs="unbounded" />
	 </xsd:sequence>
 </xsd:complexType>
       <xsd:complexType name="tr">
        <xsd:attribute name="id" type="xsd:int" use="required"/>
        <xsd:attribute name="testcasecount" type="xsd:int" use="optional"/>
        <xsd:attribute name="result" type="xsd:string" use="required"/>
        <xsd:attribute name="total" type="xsd:int" use="optional"/>
        <xsd:attribute name="passed" type="xsd:int" use="required"/>
        <xsd:attribute name="failed" type="xsd:int" use="required"/>
        <xsd:attribute name="inconclusive" type="xsd:int" use="required"/>
        <xsd:attribute name="skipped" type="xsd:int" use="required"/>
        <xsd:attribute name="asserts" type="xsd:int" use="required"/>
        <xsd:attribute name="start-time" type="xsd:date" use="required"/>
        <xsd:attribute name="end-time" type="xsd:date" use="required"/>
        <xsd:attribute name="duration" type="xsd:decimal" use="required"/>
    </xsd:complexType> 
            <xsd:element name="Environment"> 
	    <xsd:complexType>
	        <xsd:attribute name="nunit-version" type="xsd:string" use="required"/>
	        <xsd:attribute name="clr-version" type="xsd:string" use="required"/>
	        <xsd:attribute name="os-version" type="xsd:string" use="required"/>
	       <xsd:attribute name="platform" type="xsd:string" use="optional"/>
	       <xsd:attribute name="cwd" type="xsd:string" use="required"/>
	       <xsd:attribute name="machine-name" type="xsd:string" use="optional"/>
	       <xsd:attribute name="user" type="xsd:string" use="required"/>
	       <xsd:attribute name="user-domain" type="xsd:string" use="required"/>
	       <xsd:attribute name="culture" type="xsd:string" use="required"/>
	       <xsd:attribute name="uiculture" type="xsd:string" use="required"/>
	       <xsd:attribute name="os-architecture" type="xsd:string" use="required"/>
	    </xsd:complexType> 
            </xsd:element>
	<xsd:element name="TestSuite" type="xsd:string"/>    
	<xsd:element name="CommandLine" type="xsd:string"/>    
	<xsd:element name="Settings" type="xsd:string"/>    
</xsd:schema>

**Comment by CharliePoole** [comment]
As you see from the milestone assigned to it, we are not working on this right now, so I'll only give a few brief comments. I don't really want to try to create an XSD until after we settle on the actual file content. I think a "friendlier" format will be better for people to comment on.

The file content is being worked as issue #254 and the current documentation of the file comment is at https://github.com/nunit/nunit/wiki/Test-Result-XML-Format We are waiting for comments!!!

Since you have gone to this trouble, however, here are a few comments to your comments. :-)

* Generating an XSD schema from the file: We are not doing that.

* Description is not part of the file because it's a property rather than an attribute.

* Use of hyphen in element names: The NUnit 2.x result format has used hyphens since the year 2000. It has never appeared to cause any problems. Is there a reason you say it is not a good practice? Changing all the elements would be a significant programming change that we would be unlikely to do at this late date.

As I explained above, issue #254 is a pre-requisite to this one. We do not want to have to work on the XSD more than once and we don't actually need it to release NUnit 3.0. That's why it's in the milestone of 3.0-Extras, that is, individual bits to be worked on after release.

**Comment by ravichandranjv** [comment]
Hi Charlie, Thanks for spending some time on XML. :) 

* Use of hyphens as not a good practice is simply in the context of the XSD element naming conventions, which follows the upper case camel convention. 

You have mentioned above, "some people like to validate against a XSD..", which is the reason why I posted the XSD. 

By having the XSD, people can use tools to generate a schema diagram from the XSD file, which can also serve as documentation, along with comments next to the schema boxes in the diagram plus, validate the XML file, which may be important for people who may want to further work through the XML using Xpath or XSL.

Having a **beta** XSD, now, however broken or incomplete, will serve its purpose in gaining some insight into the type of XML that should be generated from code. 

For instance, in issue [#254](https://github.com/nunit/nunit/issues/254), the issue you have highlighted under `<environment>` gets resolved if we approach the XML creation through XSD standards, which says that if you wish to have multiple values in a data set for the same attribute/property of an object then use elements, so the two different or more runtime versions can be accommodated under `<environment>` itself rather than at the assembly level, which should gain  favor because it is an OO approach, as per your stated preference there -.to have a semi-colon separated set of values under a single field.

The reason I mention it here and not in the comments of [#254](https://github.com/nunit/nunit/issues/254) is because it is related more to XSD (as and when it may get created) because values, too, can be/get constrained in XSD and if ignored, can cause errors at the time of validation, post-XML generation causing problems for report generation of the test results. 

So, even if it is not priority now, having a glimpse of the kind of XSD that may be needed later helps so as to reduce any chance of a nasty code change just so that a schema validation error that has nothing to with the NUnit user, causes the tests to fail as in issue [850](https://github.com/nunit/nunit/issues/850).

**Comment by CharliePoole** [comment]
Thanks for your comments. We will keep them in mind when we work on the
XSD. NUnit itself does not require or make any use of an XSD, so it is an
extra that we will provide for those who need it only after the release.

We have been working on the new format for the result file for over a year
and it is already being created. The documentation I created reflects what
we are actually doing. It's important to us to have this reviewed. I
understand that you are suggesting a different process, but we have a
process already under way. We are stretched too thin to add any new tasks
right now. If you would like to help, please stick to the approach we are
taking and make comments either on issue #254 or through the developer list.

**Comment by ravichandranjv** [comment]
Hi Charlie,

I am not suggesting any process change nor proposing any new tasks just highlighted some points related to XSD. 

I am happy to go by your judgement/direction.

**Comment by stashell** [comment]
Hello everyone. We are writing tool which will parse also NUnit3 test results and thus I have a few questions regarding the XSD scheme and stability of the test-results XML format design.
- Any news about the "existence" of XSD for NUnit3 test-results ?
- How stable is the XML test-results design currently ? 
- Do you expect many changes in the further 3.x versions ?    

Thank you.

Patrik

**Comment by rprouse** [comment]
@stashell there is currently no XSD for the test-results format, but the design is fairly stable. The XML format is part of the API between the NUnit Engine and runners including 3rd Party runners, so we are careful to not change it unless there are bugs. Any changes tend to be minor and additive. For example, we might add a new attribute or occasionally a new child element, but that is rare.

To get a good example of all of the combinations, I would suggest creating a mock test suite that has a variety of passing, failing, ignored, explicit and errors. Errors are tests that throw non-NUnit assert exceptions. I would also use properties and categories on tests and include test output using the `TestContext.Write` methods.

**Comment by CharliePoole** [comment]
@stashell In addition, read the docs! https://github.com/nunit/docs/wiki/Test-Result-XML-Format

@rprouse Maybe we should try to get this done. I don't think either of us is very enthusiatic about the job. :-) Any XML lovers out there to volunteer?

**Comment by CharliePoole** [comment]
Issue moved to [nunit/nunit-console #13](https://github.com/nunit/nunit-console/issues/13) via [**ZenHub**](https://www.zenhub.com/)
